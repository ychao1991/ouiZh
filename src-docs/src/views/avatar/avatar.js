/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiAvatar, OuiSpacer, OuiTitle } from '../../../../src/components';

export default () => (
  <div>
    <OuiAvatar size="s" name="Raphael" />
    &emsp;
    <OuiAvatar size="m" name="Donatello" />
    &emsp;
    <OuiAvatar size="l" name="Leonardo" color="#BD10E0" />
    &emsp;
    <OuiAvatar size="xl" name="Michelangelo" />
    <OuiSpacer />
    <OuiTitle size="xs">
      <h2>With image</h2>
    </OuiTitle>
    <OuiSpacer />
    <OuiAvatar
      size="s"
      name="Cat"
      imageUrl="https://source.unsplash.com/64x64/?cat"
    />
    &emsp;
    <OuiAvatar
      size="m"
      name="Cat"
      imageUrl="https://source.unsplash.com/64x64/?cat"
    />
    &emsp;
    <OuiAvatar
      size="l"
      name="Cat"
      imageUrl="https://source.unsplash.com/64x64/?cat"
    />
    &emsp;
    <OuiAvatar
      size="xl"
      name="Cat"
      imageUrl="https://source.unsplash.com/64x64/?cat"
    />
  </div>
);
