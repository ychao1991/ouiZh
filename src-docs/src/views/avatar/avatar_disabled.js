/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';

import { OuiAvatar } from '../../../../src/components';

export default () => (
  <div>
    <OuiAvatar
      size="m"
      type="space"
      name="Disabled"
      initials="Di"
      initialsLength={2}
      isDisabled={true}
    />
    &emsp;
    <OuiAvatar
      size="m"
      type="user"
      name="User"
      initials="En"
      initialsLength={2}
      isDisabled={true}
    />
    &emsp;
    <OuiAvatar
      size="m"
      name="Cat"
      imageUrl="https://source.unsplash.com/64x64/?cat"
      isDisabled={true}
    />
    &emsp;
    <OuiAvatar
      size="m"
      name="Management"
      iconType="managementApp"
      isDisabled={true}
    />
  </div>
);
