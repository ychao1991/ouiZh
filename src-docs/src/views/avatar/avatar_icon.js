/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React, { useContext } from 'react';
import { ThemeContext } from '../../components';

import { OuiAvatar, OuiSpacer, OuiTitle } from '../../../../src/components';

export default () => {
  const themeContext = useContext(ThemeContext);

  /**
   * Setup theme based on current light/dark theme
   */
  const isDarkTheme = themeContext.theme.includes('dark');

  return (
    <div>
      <OuiTitle size="xs">
        <h2>Avatar colors and sizes</h2>
      </OuiTitle>
      <OuiSpacer />
      <OuiAvatar size="s" name="Small size" iconType="managementApp" />
      &emsp;
      <OuiAvatar size="m" name="Medium size" iconType="managementApp" />
      &emsp;
      <OuiAvatar size="l" name="Large" iconType="managementApp" />
      &emsp;
      <OuiAvatar
        size="xl"
        color="plain"
        name="Plain color"
        iconType="managementApp"
      />
      <OuiSpacer />
      <OuiTitle size="xs">
        <h2>Icon colors and sizes</h2>
      </OuiTitle>
      <OuiSpacer />
      <OuiAvatar name="Avatar color" iconType="managementApp" color="#BD10E0" />
      &emsp;
      <OuiAvatar
        name="Custom iconColor"
        iconType="managementApp"
        color={isDarkTheme ? '#103148' : '#E6F1FA'}
        iconColor="primary"
      />
      &emsp;
      <OuiAvatar
        name="Null iconColor"
        iconType="managementApp"
        color={isDarkTheme ? '#343741' : '#D3DAE6'}
        iconColor={null}
      />
      &emsp;
      <OuiAvatar name="Large iconSize" iconType="managementApp" iconSize="l" />
      &emsp;
    </div>
  );
};
