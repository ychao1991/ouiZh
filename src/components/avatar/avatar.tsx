/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { HTMLAttributes, FunctionComponent, CSSProperties } from 'react';
import { CommonProps, ExclusiveUnion, keysOf } from '../common';
import classNames from 'classnames';

import { isColorDark, hexToRgb, isValidHex } from '../../services/color';
import { ouiPaletteColorBlindBehindText, toInitials } from '../../services';
import { IconType, OuiIcon, IconSize, IconColor } from '../icon';

const sizeToClassNameMap = {
  s: 'ouiAvatar--s',
  m: 'ouiAvatar--m',
  l: 'ouiAvatar--l',
  xl: 'ouiAvatar--xl',
};

export const SIZES = keysOf(sizeToClassNameMap);
export type OuiAvatarSize = keyof typeof sizeToClassNameMap;

const typeToClassNameMap = {
  space: 'ouiAvatar--space',
  user: 'ouiAvatar--user',
};

export const TYPES = keysOf(typeToClassNameMap);
export type OuiAvatarType = keyof typeof typeToClassNameMap;

/**
 * The avatar can only display one type of content,
 * initials, or image, or iconType
 */
type _OuiAvatarContent = ExclusiveUnion<
  ExclusiveUnion<
    {
      /**
       * Custom initials (max 2 characters).
       * By default will take the first character (of each word).
       */
      initials?: string;

      /**
       * Specify how many characters to show (1 or 2).
       * By default, will show based on number of words (max first 2).
       */
      initialsLength?: 1 | 2;
    },
    {
      /**
       * Path to an image to display instead of initials
       */
      imageUrl: string;
    }
  >,
  {
    /**
     * Any OUI glyph, logo or custom icon to display instead of initials
     */
    iconType: IconType;
    /**
     * Manually change icon size
     */
    iconSize?: IconSize;
    /**
     * Manually change icon color
     */
    iconColor?: IconColor | null;
  }
>;

export type OuiAvatarProps = Omit<HTMLAttributes<HTMLDivElement>, 'color'> &
  CommonProps &
  _OuiAvatarContent & {
    /**
     * Full name of avatar for title attribute and calculating initial if not provided
     */
    name: string;

    /**
     * Accepts hex values like `#FFFFFF`, `#000` otherwise a viz palette color will be assigned.
     * Or pass `'plain'` for an empty shade or `null` to remove entirely and the text/icon color will `inherit`
     */
    color?: string | 'plain' | null;

    /**
     * The type of avatar mainly controlling the shape.
     * `user` = circle
     * `space` = rounded square
     */
    type?: OuiAvatarType;
    size?: OuiAvatarSize;

    /**
     * Grays out the avatar to simulate being disabled
     */
    isDisabled?: boolean;
  };

export const OuiAvatar: FunctionComponent<OuiAvatarProps> = ({
  className,
  color,
  imageUrl,
  initials,
  initialsLength,
  iconType,
  iconSize,
  iconColor,
  name,
  size = 'm',
  type = 'user',
  isDisabled = false,
  style,
  ...rest
}) => {
  const visColors = ouiPaletteColorBlindBehindText();

  const classes = classNames(
    'ouiAvatar',
    sizeToClassNameMap[size],
    typeToClassNameMap[type],
    {
      'ouiAvatar-isDisabled': isDisabled,
      'ouiAvatar--plain': color === 'plain',
    },
    className
  );

  checkValidInitials(initials);

  const avatarStyle: CSSProperties = style || {};
  let iconCustomColor = iconColor;

  const isNamedColor = color === 'plain' || color === null;
  if (!isNamedColor) {
    checkValidColor(color);

    const assignedColor =
      color || visColors[Math.floor(name.length % visColors.length)];
    const textColor = isColorDark(...hexToRgb(assignedColor))
      ? '#FFFFFF'
      : '#000000';

    avatarStyle.backgroundColor = assignedColor;
    avatarStyle.color = textColor;

    // Allow consumers to let the icons keep their default color (like app icons)
    // when passing `iconColor = null`, otherwise continue to pass on `iconColor` or adjust with textColor
    iconCustomColor = iconColor || iconColor === null ? iconColor : textColor;
  }

  if (imageUrl) {
    avatarStyle.backgroundImage = `url(${imageUrl})`;
  }

  let content;
  if (!imageUrl && !iconType) {
    // Create the initials
    const calculatedInitials = toInitials(name, initialsLength, initials);
    content = <span aria-hidden="true">{calculatedInitials}</span>;
  } else if (iconType) {
    content = (
      <OuiIcon
        className="ouiAvatar__icon"
        size={iconSize || size}
        type={iconType}
        aria-label={name}
        color={iconCustomColor === null ? undefined : iconCustomColor}
      />
    );
  }

  return (
    <div
      className={classes}
      style={avatarStyle}
      aria-label={name}
      title={name}
      {...rest}>
      {content}
    </div>
  );
};

// TODO: Migrate to a service
export const checkValidColor = (color: OuiAvatarProps['color']) => {
  const validHex = (color && isValidHex(color)) || color === 'plain';
  if (color && !validHex) {
    throw new Error(
      'OuiAvatar needs to pass a valid color. This can either be a three ' +
        'or six character hex value'
    );
  }
};

function checkValidInitials(initials: OuiAvatarProps['initials']) {
  // Must be a string of 1 or 2 characters
  if (initials && initials.length > 2) {
    console.warn(
      'OuiAvatar only accepts a max of 2 characters for the initials as a string. It is displaying only the first 2 characters.'
    );
  }
}
