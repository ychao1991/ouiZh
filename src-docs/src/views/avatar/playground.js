/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { PropTypes } from 'react-view';
import { OuiAvatar, checkValidColor } from '../../../../src/components/avatar';
import {
  propUtilityForPlayground,
  iconValidator,
} from '../../services/playground';

export default () => {
  const docgenInfo = Array.isArray(OuiAvatar.__docgenInfo)
    ? OuiAvatar.__docgenInfo[0]
    : OuiAvatar.__docgenInfo;
  const propsToUse = propUtilityForPlayground(docgenInfo.props);
  propsToUse.name.value = 'Avatar';

  propsToUse.color = {
    ...propsToUse.color,
    value: undefined,
    custom: {
      ...propsToUse.color.custom,
      validator: (val) => {
        try {
          checkValidColor(val);
          return true;
        } catch (error) {
          return false;
        }
      },
    },
  };

  propsToUse.initialsLength = {
    ...propsToUse.initialsLength,
    type: PropTypes.Number,
  };

  propsToUse.iconType = iconValidator(propsToUse.iconType);

  return {
    config: {
      componentName: 'OuiAvatar',
      props: propsToUse,
      scope: {
        OuiAvatar,
      },
      imports: {
        '@opensearch-project/oui': {
          named: ['OuiAvatar'],
        },
      },
    },
  };
};
