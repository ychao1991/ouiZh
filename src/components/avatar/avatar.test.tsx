/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render } from 'enzyme';
import { requiredProps } from '../../test/required_props';

import { OuiAvatar, SIZES } from './avatar';

describe('OuiAvatar', () => {
  test('is rendered', () => {
    const component = render(<OuiAvatar name="name" {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  test('allows a name composed entirely of whitespace', () => {
    const component = render(<OuiAvatar name="  " {...requiredProps} />);

    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    describe('imageUrl', () => {
      it('is rendered', () => {
        const component = render(
          <OuiAvatar name="name" imageUrl="image url" />
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('iconType', () => {
      it('is rendered', () => {
        const component = render(<OuiAvatar name="name" iconType="bolt" />);

        expect(component).toMatchSnapshot();
      });

      it('and iconSize is rendered', () => {
        const component = render(
          <OuiAvatar name="name" iconType="bolt" iconSize="xl" />
        );

        expect(component).toMatchSnapshot();
      });

      it('and iconColor is rendered', () => {
        const component = render(
          <OuiAvatar name="name" iconType="bolt" iconColor="primary" />
        );

        expect(component).toMatchSnapshot();
      });

      it('and iconColor as null is rendered', () => {
        const component = render(
          <OuiAvatar name="name" iconType="bolt" iconColor={null} />
        );

        expect(component).toMatchSnapshot();
      });
    });

    describe('size', () => {
      SIZES.forEach((size) => {
        it(`${size} is rendered`, () => {
          const component = render(<OuiAvatar name="name" size={size} />);

          expect(component).toMatchSnapshot();
        });
      });
    });

    describe('initials', () => {
      it('is rendered', () => {
        const component = render(<OuiAvatar name="name" initials="lo" />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('initialsLength', () => {
      it('is rendered', () => {
        const component = render(<OuiAvatar name="name" initialsLength={2} />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('type', () => {
      it('is rendered', () => {
        const component = render(<OuiAvatar name="name" type="space" />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('color', () => {
      it('as string is rendered', () => {
        const component = render(<OuiAvatar name="name" color="#000" />);

        expect(component).toMatchSnapshot();
      });

      it('as null is rendered', () => {
        const component = render(<OuiAvatar name="name" color={null} />);

        expect(component).toMatchSnapshot();
      });

      it('as plain is rendered', () => {
        const component = render(<OuiAvatar name="name" color="plain" />);

        expect(component).toMatchSnapshot();
      });
    });

    describe('isDisabled', () => {
      it('is rendered', () => {
        const component = render(<OuiAvatar name="name" isDisabled={true} />);

        expect(component).toMatchSnapshot();
      });
    });
  });

  test('should throw error if color is not a hex', () => {
    const component = () =>
      render(<OuiAvatar name="name" color="rgba(0,0,0,0)" />);

    expect(component).toThrowErrorMatchingSnapshot();
  });
});
