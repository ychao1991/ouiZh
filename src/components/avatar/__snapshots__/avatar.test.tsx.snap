// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiAvatar allows a name composed entirely of whitespace 1`] = `
<div
  aria-label="aria-label"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user testClass1 testClass2"
  data-test-subj="test subject string"
  style="background-color:#ee789d;color:#000000"
  title="  "
>
  <span
    aria-hidden="true"
  >
      
  </span>
</div>
`;

exports[`OuiAvatar is rendered 1`] = `
<div
  aria-label="aria-label"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user testClass1 testClass2"
  data-test-subj="test subject string"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props color as null is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props color as plain is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user ouiAvatar--plain"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props color as string is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#000;color:#FFFFFF"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props iconType and iconColor as null is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-label="name"
    class="ouiAvatar__icon"
    data-ouiicon-type="bolt"
  />
</div>
`;

exports[`OuiAvatar props iconType and iconColor is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-label="name"
    class="ouiAvatar__icon"
    color="primary"
    data-ouiicon-type="bolt"
  />
</div>
`;

exports[`OuiAvatar props iconType and iconSize is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-label="name"
    class="ouiAvatar__icon"
    color="#000000"
    data-ouiicon-type="bolt"
  />
</div>
`;

exports[`OuiAvatar props iconType is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-label="name"
    class="ouiAvatar__icon"
    color="#000000"
    data-ouiicon-type="bolt"
  />
</div>
`;

exports[`OuiAvatar props imageUrl is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000;background-image:url(image url)"
  title="name"
/>
`;

exports[`OuiAvatar props initials is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    l
  </span>
</div>
`;

exports[`OuiAvatar props initialsLength is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    na
  </span>
</div>
`;

exports[`OuiAvatar props isDisabled is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user ouiAvatar-isDisabled"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props size l is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--l ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props size m is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props size s is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--s ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props size xl is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--xl ouiAvatar--user"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar props type is rendered 1`] = `
<div
  aria-label="name"
  class="ouiAvatar ouiAvatar--m ouiAvatar--space"
  style="background-color:#e4a6c7;color:#000000"
  title="name"
>
  <span
    aria-hidden="true"
  >
    n
  </span>
</div>
`;

exports[`OuiAvatar should throw error if color is not a hex 1`] = `"OuiAvatar needs to pass a valid color. This can either be a three or six character hex value"`;
